server/src/main/java/io/druid/metadata/SQLMetadataSupervisorManager.java

```diff
@@ -0,0 +1,250 @@
+                          throws SQLException
+                      {
+                        try {
+                          SupervisorSpec payload = jsonMapper.readValue(
+                              r.getBytes("payload"),
+                              new TypeReference<SupervisorSpec>()
+                              {
+                              }
+                          );
+                          return Pair.of(
+                              r.getString("spec_id"),
+                              new VersionedSupervisorSpec(payload, r.getString("version"))
+                          );
+                        }
+                        catch (IOException e) {
+                          throw Throwables.propagate(e);
+                        }
+                      }
+                    }
+                ).fold(
+                    Maps.<String, List<VersionedSupervisorSpec>>newHashMap(),
+                    new Folder3<Map<String, List<VersionedSupervisorSpec>>, Pair<String, VersionedSupervisorSpec>>()
+                    {
+                      @Override
+                      public Map<String, List<VersionedSupervisorSpec>> fold(
+                          Map<String, List<VersionedSupervisorSpec>> retVal,
+                          Pair<String, VersionedSupervisorSpec> pair,
+                          FoldController foldController,
+                          StatementContext statementContext
+                      ) throws SQLException
+                      {
+                        try {
+                          String specId = pair.lhs;
+                          if (!retVal.containsKey(specId)) {
+                            retVal.put(specId, Lists.<VersionedSupervisorSpec>newArrayList());
+                          }
+
+                          retVal.get(specId).add(pair.rhs);
+                          return retVal;
+                        }
+                        catch (Exception e) {
+                          throw Throwables.propagate(e);
+                        }
+                      }
+                    }
+                );
+              }
+            }
+        )
+    );
+  }
+
+  @Override
+  public Map<String, SupervisorSpec> getLatest()
+  {
+    return ImmutableMap.copyOf(
+        dbi.withHandle(
+            new HandleCallback<Map<String, SupervisorSpec>>()
+            {
+              @Override
+              public Map<String, SupervisorSpec> withHandle(Handle handle) throws Exception
+              {
+                return handle.createQuery(
+                    String.format(
+                        "SELECT r.spec_id, r.payload "
+                        + "FROM %1$s r "
+                        + "INNER JOIN(SELECT spec_id, max(version) as version FROM %1$s GROUP BY spec_id) latest "
+                        + "ON r.spec_id = latest.spec_id and r.version = latest.version",
+                        getSupervisorsTable()
```

Comment:
limiting items in history may potentially help the performance of this query too.